presidio-analyzer/presidio_analyzer/nlp_engine/spacy_nlp_engine.py

```diff
@@ -76,20 +138,99 @@ def get_nlp(self, language: str) -> Language:
-        entities = doc.ents
+
+        entities = self._get_entities(doc)
+        scores = self._get_scores_for_entities(doc)
+
+        entities, scores = self._get_updated_entities(entities, scores)
+
         return NlpArtifacts(
             entities=entities,
             tokens=doc,
             tokens_indices=tokens_indices,
             lemmas=lemmas,
             nlp_engine=self,
             language=language,
+            scores=scores,
         )
+
+    def _get_entities(self, doc: Doc) -> List[Span]:
+        """
+        Extract entities out of a spaCy pipeline, depending on the type of pipeline.
+
+        For normal spaCy, this would be doc.ents
+        :param doc: the output spaCy doc.
+        :return: List of entities
+        """
+
+        return doc.ents
+
+    def _get_scores_for_entities(self, doc: Doc) -> List[float]:
+        """Extract scores for entities from the doc.
+
+        Since spaCy does not provide confidence scores for entities by default,
+        we use the default score from the ner model configuration.
+        :param doc: SpaCy doc
+        """
+
+        entities = doc.ents
+        scores = [self.ner_model_configuration.default_score] * len(entities)
+        return scores
+
+    def _get_updated_entities(
+        self, entities: List[Span], scores: List[float]
+    ) -> Tuple[List[Span], List[float]]:
+        """
+        Get an updated list of entities based on the ner model configuration.
+
+        Remove entities that are in labels_to_ignore,
+        update entity names based on model_to_presidio_entity_mapping
+
+        :param entities: Entities that were extracted from a spaCy pipeline
+        :param scores: Original confidence scores for the entities extracted
+        :return: Tuple holding the entities and confidence scores
+        """
+        if len(entities) != len(scores):
+            raise ValueError("Entities and scores must be the same length")
+
+        new_entities = []
+        new_scores = []
+
+        mapping = self.ner_model_configuration.model_to_presidio_entity_mapping
+        to_ignore = self.ner_model_configuration.labels_to_ignore
+        for ent, score in zip(entities, scores):
+            # Remove model labels in the ignore list
+            if ent.label_ in to_ignore:
+                continue
+
+            # Update entity label based on mapping
+            if ent.label_ in mapping:
+                ent.label_ = mapping[ent.label_]
+            else:
+                logger.warning(
+                    f"Entity {ent.label_} is not mapped to a Presidio entity, "
+                    f"but keeping anyway. "
+                    f"Add to `NerModelConfiguration.labels_to_ignore` to remove."
+                )
+
+            # Remove presidio entities in the ignore list
+            if ent.label_ in to_ignore:
+                continue
+
+            new_entities.append(ent)
+
+            # Update score if entity is in low score entity names
```

Comment:
Can the score ever be updated to be higher than the default value?

Reply:
There's no limitation here. Do you think there should be?

Reply:
Looking at how scores are handled in this file, it seems that scores are initially assigned the default value and then only the scores associated with certain entities are updated, namely, by being multiplied by the low confidence score multiplier.

However, from a quick search, I now see that spacy NER might not return confidence scores. To clarify, is this the reason for this logic?

Reply:
The intention is to allow users to handle entities that have more errors on their data. Models either return a fixed confidence value (like 0.85) or the softmax output (on transformers based models). Still, there are cases where different entities have higher error rates.

Reply:
I would be happy to consider other approaches to this. There is definitely a requirement to handle different entities differently.

Reply:
Got it. In this case, I am okay with leaving it as-is, but we can revisit in the future as needed.